src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPECompilerManager.cs

```diff
@@ -0,0 +1,372 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+
+using Microsoft.VisualStudio.IO;
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.LanguageServices;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.ProjectSystem.Utilities;
+using Microsoft.VisualStudio.Threading;
+using Microsoft.VisualStudio.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.TempPE
+{
+    internal class TempPECompilerManager : OnceInitializedOnceDisposedAsync
+    {
+        private static readonly TimeSpan s_compilationDelayTime = TimeSpan.FromMilliseconds(500);
+
+        private readonly ConfiguredProject _project;
+        private readonly IProjectSubscriptionService _projectSubscriptionService;
+        private readonly IActiveWorkspaceProjectContextHost _activeWorkspaceProjectContextHost;
+        private readonly IDesignTimeInputsDataSource _inputsDataSource;
+        private readonly IDesignTimeInputsFileWatcher _fileWatcher;
+        private readonly ITempPECompiler _compiler;
+        private readonly IFileSystem _fileSystem;
+        private readonly TaskDelayScheduler _scheduler;
+
+        private readonly DisposableBag _disposables = new DisposableBag();
+
+        private ImmutableArray<string> _designTimeInputs;
+        private ImmutableArray<string> _sharedDesignTimeInputs;
+        private ImmutableDictionary<string, bool> _filesToCompile = ImmutableDictionary<string, bool>.Empty.WithComparers(StringComparers.Paths); // Key is filename, value is whether to ignore the last write time check
+        private string? _outputPath;
+        private readonly TaskCompletionSource<bool> _receivedProjectRuleSource = new TaskCompletionSource<bool>();
+
+        private ITargetBlock<IProjectVersionedValue<Tuple<DesignTimeInputs, IProjectSubscriptionUpdate>>>? _inputsActionBlock;
+        private ITargetBlock<IProjectVersionedValue<string[]>>? _fileWatcherActionBlock;
+
+        [ImportingConstructor]
+        public TempPECompilerManager(ConfiguredProject project,
```

Comment:
As I was reading the PR, I found it a little hard to keep track of when the mutable state of this class is assigned and can be accessed. I would consider merging it into a single field representing your "current", that is assigned all at once atomically.